File: code/C/第15章 动态规划/15.2 硬币找零/coins_recursive.c
#include <stdio.h>
#include <limits.h>

int coins[] = {2, 5, 7};
int n = sizeof(coins) / sizeof(coins[0]);

int min(int a, int b) {
    return a < b ? a : b;
}

int getMinCoins(int price) {
    // 0元钱只需要0枚硬币
    if(price == 0) {
        return 0;
    }
    // 初始化为无穷大
    int coinNum = INT_MAX - 1;
    // 最后一枚硬币是2元
    if(price >= 2) {
        coinNum = min(getMinCoins(price-2) + 1, coinNum);
    }
    // 最后一枚硬币是5元
    if(price >= 5) {
        coinNum = min(getMinCoins(price-5) + 1, coinNum);
    }
    // 最后一枚硬币是7元
    if(price >= 7) {
        coinNum = min(getMinCoins(price-7) + 1, coinNum);
    }
    return coinNum;
}

int main() {
    int price = 27;
    printf("%d\n", getMinCoins(price));
    return 0;
}

File: code/Python/Chapter12/12-2/coins_recursive.py
COINS = [2, 5, 7]
INF = 2147483647

def get_min_coins(price):
    # 0元钱只需要0枚硬币
    if price == 0:
        return 0
    # 初始化为无穷大
    coin_num = INF
    # 最后一枚硬币是2元
    if price >= 2:
        coin_num = min(get_min_coins(price-2) + 1, coin_num)
    # 最后一枚硬币是5元
    if price >= 5:
        coin_num = min(get_min_coins(price-5) + 1, coin_num)
    # 最后一枚硬币是7元
    if price >= 7:
        coin_num = min(get_min_coins(price-7) + 1, coin_num)
    return coin_num

def main():
    price = 27
    print(get_min_coins(price))

if __name__ == "__main__":
    main()